pkg/controller/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,475 @@
+			for _, v := range devices {
+				deviceInfo = append(deviceInfo, vGPUMap[v])
+			}
+			gpuMap[device.DeviceName] = deviceInfo
+		}
+	}
+
+	// generate allocation details
+	deviceDetails := generateAllocationDetails(hostDeviceMap, gpuMap)
+	deviceDetailsBytes, err := json.Marshal(deviceDetails)
+	if err != nil {
+		return fmt.Errorf("error marshalling deviceDetails: %v", err)
+	}
+
+	return h.reconcileVMResourceAllocationAnnotation(vmi, string(deviceDetailsBytes))
+}
+
+// findPodForVMI leverages the fact that each pod associated with a VMI a label vm.kubevirt.io/name: $vmName
+// this makes it easier to find the correct pod
+func (h *Handler) findPodForVMI(vmi *kubevirtv1.VirtualMachineInstance) (*corev1.Pod, error) {
+	podList, err := h.pod.List(vmi.Namespace, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", kubevirtVMLabelKey, vmi.Name),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing pods: %v", err)
+	}
+
+	// if more than 1 pod is returned make sure only 1 is running
+	// if more than 1 is running then error out since there may be a migration
+	// running and reconcile again
+	var runningPod corev1.Pod
+
+	var count int
+	for _, pod := range podList.Items {
+		if pod.Status.Phase == corev1.PodRunning {
+			runningPod = pod // we copy pod in case there is only 1 running, avoids having to iterate again
+			count++
+		}
+	}
+
+	if count != 1 {
+		return nil, fmt.Errorf("expected to find 1 pod, but found %d associated with vmi %s", count, vmi.Name)
+	}
+
+	return &runningPod, nil
+}
+
+func (h *Handler) getPodEnv(pod *corev1.Pod) ([]byte, error) {
+	e, err := executor.NewRemoteCommandExecutor(h.ctx, h.config, pod)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up remote executor for pod %s: %v", pod.Name, err)
+	}
+	return e.Run("env", []string{})
+}
+
+// convertEnvToMap converts env info to a map to make it easier to find device
+// allocation details
+func convertEnvToMap(podEnv []byte) (map[string]string, error) {
+	var contents []string
+	bufReader := bufio.NewReader(bytes.NewReader(podEnv))
+
+	for {
+		line, err := bufReader.ReadBytes('\n')
+		if len(line) != 0 {
+			contents = append(contents, strings.TrimSuffix(string(line), "\n"))
+		}
+
+		if err != nil {
+			if err != io.EOF {
+				return nil, err
+			}
+			break
+		}
+	}
+
+	resultMap := make(map[string]string)
+	for _, line := range contents {
+		lineArr := strings.Split(line, "=")
+		resultMap[lineArr[0]] = strings.Join(lineArr[1:], "=")
+	}
+	return resultMap, nil
+}
+
+// check and clear deviceAllocation annotations if needed
+func (h *Handler) checkAndClearDeviceAllocation(vmi *kubevirtv1.VirtualMachineInstance) error {
+	vmObj, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return fmt.Errorf("error fetching vm %s from cache: %s", vmi.Name, err)
+	}
+	if vmObj.Annotations == nil {
+		return nil
+	}
+
+	_, ok := vmObj.Annotations[v1beta1.DeviceAllocationKey]
+	// no key, nothing is needed
+	if !ok {
+		return nil
+	}
```

Comment:
NIT: Reading a nil map is allowed, we could remove it because we don't update the map later in this function.